File: algorithms/rust/src/sort_and_search.rs
use std::cell::RefCell;
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::rc::Rc;

/// Linked list node using Rc<RefCell> for easier split/merge operations
#[derive(Debug, PartialEq)]
pub struct Node {
    pub val: i32,
    pub next: Option<Rc<RefCell<Node>>>,
}

pub type NodeRef = Rc<RefCell<Node>>;

impl Node {
    pub fn new(val: i32) -> NodeRef {
        Rc::new(RefCell::new(Node { val, next: None }))
    }

    pub fn from_vec(values: &[i32]) -> Option<NodeRef> {
        if values.is_empty() {
            return None;
        }
        let head = Node::new(values[0]);
        let mut curr = Rc::clone(&head);
        for &val in &values[1..] {
            let new_node = Node::new(val);
            curr.borrow_mut().next = Some(Rc::clone(&new_node));
            curr = new_node;
        }
        Some(head)
    }

    pub fn to_vec(head: &Option<NodeRef>) -> Vec<i32> {
        let mut result = Vec::new();
        let mut curr = head.clone();
        while let Some(node) = curr {
            result.push(node.borrow().val);
            curr = node.borrow().next.clone();
        }
        result
    }
}

/// Sort Linked List
///
/// Given the head of a singly linked list, sort the linked list in ascending order.
/// Uses merge sort: split in half, sort each half, merge.
///
/// # Example
///
/// ```text
/// Input: 4 -> 2 -> 1 -> 3
///
/// Output: 1 -> 2 -> 3 -> 4
/// ```
pub fn sort_list(head: Option<NodeRef>) -> Option<NodeRef> {
    // Base case: empty or single node
    let head = head?;
    if head.borrow().next.is_none() {
        return Some(head);
    }

    // Split into two halves
    let right = split(Rc::clone(&head));
    let left = sort_list(Some(head));
    let right = sort_list(right);

    // Merge sorted halves
    merge(left, right)
}

fn split(head: NodeRef) -> Option<NodeRef> {
    let mut slow = Rc::clone(&head);
    let mut fast = Rc::clone(&head);

    loop {
        let Some(next) = fast.borrow().next.clone() else { break };
        let Some(next_next) = next.borrow().next.clone() else { break };

        let next_slow = slow.borrow().next.clone().unwrap();
        slow = next_slow;
        fast = next_next;
    }

    slow.borrow_mut().next.take()
}

fn merge(mut l1: Option<NodeRef>, mut l2: Option<NodeRef>) -> Option<NodeRef> {
    // Pick the smaller head as our result head
    let head = match (&l1, &l2) {
        (None, _) => return l2,
        (_, None) => return l1,
        (Some(a), Some(b)) => {
            if a.borrow().val <= b.borrow().val {
                let node = l1.take().unwrap();
                l1 = node.borrow().next.clone();
                node
            } else {
                let node = l2.take().unwrap();
                l2 = node.borrow().next.clone();
                node
            }
        }
    };

    let mut tail = Rc::clone(&head);

    while l1.is_some() || l2.is_some() {
        let take_l1 = match (&l1, &l2) {
            (None, _) => false,
            (_, None) => true,
            (Some(a), Some(b)) => a.borrow().val <= b.borrow().val,
        };

        let node = if take_l1 {
            let node = l1.take().unwrap();
            l1 = node.borrow().next.clone();
            node
        } else {
            let node = l2.take().unwrap();
            l2 = node.borrow().next.clone();
            node
        };

        tail.borrow_mut().next = Some(Rc::clone(&node));
        tail = node;
    }

    Some(head)
}

/// Kth Largest Integer
///
/// Return the kth largest integer in an array.
///
/// # Example
///
/// ```text
/// Input: nums = [5, 2, 4, 3, 1, 6], k = 3
///
/// Output: 4
/// ```
pub fn kth_largest(nums: &[i32], k: usize) -> i32 {
    let mut min_heap: BinaryHeap<Reverse<i32>> = BinaryHeap::new();

    for &num in nums {
        if min_heap.len() < k {
            min_heap.push(Reverse(num));
        } else if num > min_heap.peek().unwrap().0 {
            min_heap.pop();
            min_heap.push(Reverse(num));
        }
    }

    min_heap.peek().unwrap().0
}

/// Sort Array
///
/// Given an integer array, sort it in ascending order.
///
/// # Example
///
/// ```text
/// Input: nums = [6, 8, 4, 2, 7, 3, 1, 5]
///
/// Output: [1, 2, 3, 4, 5, 6, 7, 8]
/// ```
// Quicksort Complexity:
//
// Time - Average O(n log n):
//   - Good pivots split array roughly in half → log(n) levels
//   - Each level does O(n) work (partition) → O(n log n) total
//
// Time - Worst O(n²):
//   - Bad pivots (always min/max) → one side has n-1 elements
//   - Creates n levels instead of log(n) → O(n²) total
//   - Happens with already sorted arrays when using last element as pivot
//
// Space - O(log n) average, O(n) worst:
//   - Recursion depth matches the number of levels
//
pub fn sort_array(nums: &mut [i32]) {
    if nums.is_empty() {
        return;
    }
    quicksort(nums, 0, nums.len() - 1);
}

fn quicksort(nums: &mut [i32], left: usize, right: usize) {
    // Base case: if the subarray has 0 or 1 element, it's already sorted
    if left >= right {
        return;
    }

    // Partition the array and retrieve the pivot index
    let pivot_index = partition(nums, left, right);

    // Recursively sort left and right parts
    if pivot_index > 0 {
        quicksort(nums, left, pivot_index - 1);
    }
    quicksort(nums, pivot_index + 1, right);
}

fn partition(nums: &mut [i32], left: usize, right: usize) -> usize {
    let pivot = nums[right];
    let mut lo = left;

    // Move all numbers less than pivot to the left
    for i in left..right {
        if nums[i] < pivot {
            nums.swap(lo, i);
            lo += 1;
        }
    }

    // Swap pivot into its correct position
    nums.swap(lo, right);
    lo
}

/// Merge Sorted Array
///
/// Merge nums1 and nums2 into a single array sorted in non-decreasing order.
/// The result is stored in nums1, which has length m + n (last n elements are 0s).
///
/// # Example
///
/// ```text
/// Input: nums1 = [1, 2, 3, 0, 0, 0], m = 3, nums2 = [2, 5, 6], n = 3
/// Output: nums1 = [1, 2, 2, 3, 5, 6]
/// ```
pub fn merge_sorted_array(nums1: &mut [i32], m: usize, nums2: &[i32], n: usize) {
    let copy_nums1 = nums1[0..m].to_vec();
    let mut p1 = 0;
    let mut p2 = 0;

    for i in 0..(m + n) {
        if p2 >= n || (p1 < m && copy_nums1[p1] <= nums2[p2]) {
            nums1[i] = copy_nums1[p1];
            p1 += 1;
        } else {
            nums1[i] = nums2[p2];
            p2 += 1;
        }
    }
}

/// H-Index
///
/// Given an array where citations[i] is the number of citations for the ith paper,
/// return the researcher's h-index: the maximum value h such that the researcher
/// has at least h papers with at least h citations each.
///
/// # Example 1
///
/// ```text
/// Input: citations = [3, 0, 6, 1, 5]
/// Output: 3
/// Explanation: 3 papers have at least 3 citations each
/// ```
///
/// # Example 2
///
/// ```text
/// Input: citations = [1, 3, 1]
/// Output: 1
/// ```
pub fn h_index(citations: &[usize]) -> usize {
    let mut sorted = citations.to_vec();
    sorted.sort();

    let n = sorted.len();
    let mut h = 0;

    for i in 0..n {
        let papers_with_at_least = n - i;
        let citation_count = sorted[i];

        if citation_count >= papers_with_at_least {
            h = h.max(papers_with_at_least);
        }
    }

    h
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sort_list_example() {
        let head = Node::from_vec(&[4, 2, 1, 3]);
        let sorted = sort_list(head);
        assert_eq!(Node::to_vec(&sorted), vec![1, 2, 3, 4]);
    }

    #[test]
    fn test_sort_list_empty() {
        let head: Option<NodeRef> = None;
        let sorted = sort_list(head);
        assert_eq!(sorted, None);
    }

    #[test]
    fn test_sort_list_single() {
        let head = Node::from_vec(&[5]);
        let sorted = sort_list(head);
        assert_eq!(Node::to_vec(&sorted), vec![5]);
    }

    #[test]
    fn test_sort_list_already_sorted() {
        let head = Node::from_vec(&[1, 2, 3, 4, 5]);
        let sorted = sort_list(head);
        assert_eq!(Node::to_vec(&sorted), vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_sort_list_reverse() {
        let head = Node::from_vec(&[5, 4, 3, 2, 1]);
        let sorted = sort_list(head);
        assert_eq!(Node::to_vec(&sorted), vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_sort_list_duplicates() {
        let head = Node::from_vec(&[3, 1, 2, 1, 3]);
        let sorted = sort_list(head);
        assert_eq!(Node::to_vec(&sorted), vec![1, 1, 2, 3, 3]);
    }

    #[test]
    fn test_sort_array_example() {
        let mut nums = [6, 8, 4, 2, 7, 3, 1, 5];
        sort_array(&mut nums);
        assert_eq!(nums, [1, 2, 3, 4, 5, 6, 7, 8]);
    }

    #[test]
    fn test_sort_array_empty() {
        let mut nums: [i32; 0] = [];
        sort_array(&mut nums);
        assert_eq!(nums, []);
    }

    #[test]
    fn test_sort_array_single() {
        let mut nums = [5];
        sort_array(&mut nums);
        assert_eq!(nums, [5]);
    }

    #[test]
    fn test_sort_array_already_sorted() {
        let mut nums = [1, 2, 3, 4, 5];
        sort_array(&mut nums);
        assert_eq!(nums, [1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_sort_array_reverse() {
        let mut nums = [5, 4, 3, 2, 1];
        sort_array(&mut nums);
        assert_eq!(nums, [1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_sort_array_duplicates() {
        let mut nums = [3, 1, 2, 1, 3];
        sort_array(&mut nums);
        assert_eq!(nums, [1, 1, 2, 3, 3]);
    }

    #[test]
    fn test_sort_array_negative() {
        let mut nums = [3, -1, 0, -5, 2];
        sort_array(&mut nums);
        assert_eq!(nums, [-5, -1, 0, 2, 3]);
    }

    #[test]
    fn test_kth_largest_example() {
        assert_eq!(kth_largest(&[5, 2, 4, 3, 1, 6], 3), 4);
    }

    #[test]
    fn test_kth_largest_first() {
        assert_eq!(kth_largest(&[5, 2, 4, 3, 1, 6], 1), 6);
    }

    #[test]
    fn test_kth_largest_last() {
        assert_eq!(kth_largest(&[5, 2, 4, 3, 1, 6], 6), 1);
    }

    #[test]
    fn test_kth_largest_single() {
        assert_eq!(kth_largest(&[42], 1), 42);
    }

    #[test]
    fn test_kth_largest_duplicates() {
        assert_eq!(kth_largest(&[3, 2, 3, 1, 2, 4, 5, 5, 6], 4), 4);
    }

    #[test]
    fn test_kth_largest_negative() {
        assert_eq!(kth_largest(&[-1, -5, 0, 3, -2], 2), 0);
    }

    #[test]
    fn test_merge_sorted_array_example() {
        let mut nums1 = vec![1, 2, 3, 0, 0, 0];
        let nums2 = vec![2, 5, 6];
        merge_sorted_array(&mut nums1, 3, &nums2, 3);
        assert_eq!(nums1, vec![1, 2, 2, 3, 5, 6]);
    }

    #[test]
    fn test_merge_sorted_array_nums2_empty() {
        let mut nums1 = vec![1, 2, 3];
        let nums2: Vec<i32> = vec![];
        merge_sorted_array(&mut nums1, 3, &nums2, 0);
        assert_eq!(nums1, vec![1, 2, 3]);
    }

    #[test]
    fn test_merge_sorted_array_nums1_empty() {
        let mut nums1 = vec![0];
        let nums2 = vec![1];
        merge_sorted_array(&mut nums1, 0, &nums2, 1);
        assert_eq!(nums1, vec![1]);
    }

    #[test]
    fn test_merge_sorted_array_interleaved() {
        let mut nums1 = vec![1, 3, 5, 0, 0, 0];
        let nums2 = vec![2, 4, 6];
        merge_sorted_array(&mut nums1, 3, &nums2, 3);
        assert_eq!(nums1, vec![1, 2, 3, 4, 5, 6]);
    }

    #[test]
    fn test_merge_sorted_array_nums2_all_smaller() {
        let mut nums1 = vec![4, 5, 6, 0, 0, 0];
        let nums2 = vec![1, 2, 3];
        merge_sorted_array(&mut nums1, 3, &nums2, 3);
        assert_eq!(nums1, vec![1, 2, 3, 4, 5, 6]);
    }

    #[test]
    fn test_h_index_example1() {
        assert_eq!(h_index(&[3, 0, 6, 1, 5]), 3);
    }

    #[test]
    fn test_h_index_example2() {
        assert_eq!(h_index(&[1, 3, 1]), 1);
    }

    #[test]
    fn test_h_index_all_zeros() {
        assert_eq!(h_index(&[0, 0, 0]), 0);
    }

    #[test]
    fn test_h_index_single_high() {
        assert_eq!(h_index(&[100]), 1);
    }

    #[test]
    fn test_h_index_single_zero() {
        assert_eq!(h_index(&[0]), 0);
    }

    #[test]
    fn test_h_index_all_same() {
        assert_eq!(h_index(&[5, 5, 5, 5, 5]), 5);
    }
}

File: algorithms/python/sort_and_search.py
"""
Sort and Search Pattern

A collection of sorting and searching algorithms.

Key algorithms:
- Quicksort: O(n log n) avg, O(n²) worst, in-place
- Merge sort: O(n log n) always, not in-place
- Heap-based selection: O(n log k) for kth element
"""


def sort_array(nums: list[int]) -> list[int]:
    """
    Sort Array (Quicksort)

    Sort an integer array in ascending order using quicksort.

    Example:
        >>> sort_array([6, 8, 4, 2, 7, 3, 1, 5])
        [1, 2, 3, 4, 5, 6, 7, 8]

    Time Complexity: O(n log n) average, O(n²) worst
    Space Complexity: O(log n) average (recursion stack)

    Quicksort steps:
        1. Pick a pivot (we use last element)
        2. Partition: move smaller elements left, larger right
        3. Recursively sort left and right partitions

    Partition example: [6,8,4,2,7,3,1,5], pivot=5

        lo marks where the next "small" element should go.
        Scan left to right, when we find something < pivot, swap it to lo.

        [6, 8, 4, 2, 7, 3, 1, 5]    pivot=5
         ^lo
         i=0: 6 >= 5? yes, skip

        [6, 8, 4, 2, 7, 3, 1, 5]
         ^lo
         i=1: 8 >= 5? yes, skip

        [6, 8, 4, 2, 7, 3, 1, 5]
         ^lo
         i=2: 4 < 5? yes! swap arr[lo] with arr[i]

        [4, 8, 6, 2, 7, 3, 1, 5]    swapped 6↔4
            ^lo
         i=3: 2 < 5? yes! swap

        [4, 2, 6, 8, 7, 3, 1, 5]    swapped 8↔2
               ^lo
         i=4: 7 >= 5? yes, skip
         i=5: 3 < 5? yes! swap

        [4, 2, 3, 8, 7, 6, 1, 5]    swapped 6↔3
                  ^lo
         i=6: 1 < 5? yes! swap

        [4, 2, 3, 1, 7, 6, 8, 5]    swapped 8↔1
                     ^lo
         Done scanning. Now put pivot at lo:

        [4, 2, 3, 1, 5, 6, 8, 7]    swapped 7↔5
         ←--<5---→   ↑  ←->=5-→
                   pivot in final position!

        Recurse on left [4,2,3,1] and right [6,8,7]
    """
    def quicksort(left: int, right: int):
        if left >= right:
            return

        # Partition and get pivot's final position
        pivot_idx = partition(left, right)

        # Recursively sort left and right of pivot
        quicksort(left, pivot_idx - 1)
        quicksort(pivot_idx + 1, right)

    def partition(left: int, right: int) -> int:
        pivot = nums[right]  # choose last element as pivot
        lo = left            # boundary for "less than pivot" section

        # Move all elements < pivot to the left
        for i in range(left, right):
            if nums[i] < pivot:
                nums[lo], nums[i] = nums[i], nums[lo]
                lo += 1

        # Put pivot in its correct position
        nums[lo], nums[right] = nums[right], nums[lo]
        return lo

    if nums:
        quicksort(0, len(nums) - 1)
    return nums


# -----------------------------------------------------------------------------
# Tests for sort_array
# -----------------------------------------------------------------------------

def test_sort_array_example():
    assert sort_array([6, 8, 4, 2, 7, 3, 1, 5]) == [1, 2, 3, 4, 5, 6, 7, 8]


def test_sort_array_empty():
    assert sort_array([]) == []


def test_sort_array_single():
    assert sort_array([5]) == [5]


def test_sort_array_already_sorted():
    assert sort_array([1, 2, 3, 4, 5]) == [1, 2, 3, 4, 5]


def test_sort_array_reverse():
    assert sort_array([5, 4, 3, 2, 1]) == [1, 2, 3, 4, 5]


def test_sort_array_duplicates():
    assert sort_array([3, 1, 2, 1, 3]) == [1, 1, 2, 3, 3]


def test_sort_array_negative():
    assert sort_array([3, -1, 0, -5, 2]) == [-5, -1, 0, 2, 3]


def merge_sorted_array(nums1: list[int], m: int, nums2: list[int], n: int) -> None:
    """
    Merge Sorted Array

    Merge nums2 into nums1. nums1 has length m+n (last n elements are 0s).
    Modify nums1 in-place.

    Example:
        >>> nums1 = [1, 2, 3, 0, 0, 0]
        >>> merge_sorted_array(nums1, 3, [2, 5, 6], 3)
        >>> nums1
        [1, 2, 2, 3, 5, 6]

    Time Complexity: O(m + n)
    Space Complexity: O(m) - copy of nums1

    Two pointer merge: compare elements from both arrays,
    pick the smaller one each time.

    Example: nums1=[1,3,5,0,0,0] m=3, nums2=[2,4,6] n=3
        copy1 = [1,3,5]
        i=0: 1 <= 2 → take 1
        i=1: 3 > 2  → take 2
        i=2: 3 <= 4 → take 3
        i=3: 5 > 4  → take 4
        i=4: 5 <= 6 → take 5
        i=5: done   → take 6
        Result: [1,2,3,4,5,6]
    """
    # Copy nums1's actual values (first m elements)
    copy1 = nums1[:m]

    p1, p2 = 0, 0  # pointers for copy1 and nums2

    for i in range(m + n):
        # Take from copy1 if: nums2 exhausted OR copy1 has smaller value
        if p2 >= n or (p1 < m and copy1[p1] <= nums2[p2]):
            nums1[i] = copy1[p1]
            p1 += 1
        else:
            nums1[i] = nums2[p2]
            p2 += 1


# -----------------------------------------------------------------------------
# Tests for merge_sorted_array
# -----------------------------------------------------------------------------

def test_merge_sorted_array_example():
    nums1 = [1, 2, 3, 0, 0, 0]
    merge_sorted_array(nums1, 3, [2, 5, 6], 3)
    assert nums1 == [1, 2, 2, 3, 5, 6]


def test_merge_sorted_array_nums2_empty():
    nums1 = [1, 2, 3]
    merge_sorted_array(nums1, 3, [], 0)
    assert nums1 == [1, 2, 3]


def test_merge_sorted_array_nums1_empty():
    nums1 = [0]
    merge_sorted_array(nums1, 0, [1], 1)
    assert nums1 == [1]


def test_merge_sorted_array_interleaved():
    nums1 = [1, 3, 5, 0, 0, 0]
    merge_sorted_array(nums1, 3, [2, 4, 6], 3)
    assert nums1 == [1, 2, 3, 4, 5, 6]


def test_merge_sorted_array_nums2_all_smaller():
    nums1 = [4, 5, 6, 0, 0, 0]
    merge_sorted_array(nums1, 3, [1, 2, 3], 3)
    assert nums1 == [1, 2, 3, 4, 5, 6]


def h_index(citations: list[int]) -> int:
    """
    H-Index

    Given an array where citations[i] is the number of citations for the ith paper,
    return the h-index: the maximum h such that at least h papers have at least
    h citations each.

    Example:
        >>> h_index([3, 0, 6, 1, 5])
        3
        # Sorted: [0, 1, 3, 5, 6]
        # 3 papers have >= 3 citations (papers with 3, 5, 6)

    Time Complexity: O(n log n) - sorting
    Space Complexity: O(n) - sorted copy

    Sort ascending then scan. At index i, there are (n - i) papers
    with at least citations[i] citations. If citations[i] >= (n - i),
    that's a valid h-index.

    Example walkthrough: [3, 0, 6, 1, 5] → sorted: [0, 1, 3, 5, 6]
        i=0: citations=0, papers_left=5 → 0 >= 5? no
        i=1: citations=1, papers_left=4 → 1 >= 4? no
        i=2: citations=3, papers_left=3 → 3 >= 3? yes! h=3
        i=3: citations=5, papers_left=2 → 5 >= 2? yes! h=max(3,2)=3
        i=4: citations=6, papers_left=1 → 6 >= 1? yes! h=max(3,1)=3
        Answer: 3
    """
    sorted_cit = sorted(citations)
    n = len(sorted_cit)
    h = 0

    for i in range(n):
        papers_with_at_least = n - i
        if sorted_cit[i] >= papers_with_at_least:
            h = max(h, papers_with_at_least)

    return h


# -----------------------------------------------------------------------------
# Tests for h_index
# -----------------------------------------------------------------------------

def test_h_index_example1():
    assert h_index([3, 0, 6, 1, 5]) == 3


def test_h_index_example2():
    assert h_index([1, 3, 1]) == 1


def test_h_index_all_zeros():
    assert h_index([0, 0, 0]) == 0


def test_h_index_single_high():
    assert h_index([100]) == 1


def test_h_index_single_zero():
    assert h_index([0]) == 0


def test_h_index_all_same():
    assert h_index([5, 5, 5, 5, 5]) == 5


if __name__ == "__main__":
    import pytest
    pytest.main([__file__, "-v"])
